---
title: 'Happy Belly Bioinformatics: an open-source resource dedicated to helping biologists utilize bioinformatics'
tags:
- bioinformatics
- tutorial
authors:
- name: Michael D. Lee
  orcid: 0000-0001-7750-9145
  affiliation: "1"
affiliations:
- name: Exobiology Branch, NASA Ames Research Center, Moffett Field, CA, USA
  index: 1
date: 06 MAY 2019
bibliography: paper.bib
---

# Summary
*Statement of need:* Our civilization's ability to generate data has far surpassed our current, collective capabilities to efficiently access and utilize it. And while the necessary educational infrastructure is trying to adapt and grow as quickly as possible, many graduate students, postdocs, faculty, and other researchers have found themselves in situations where self-education with little direction is the only option available. This can make it very difficult to get through the initial steep learning curve of beginning to use bioinformatics effectively – particularly when this is expected to happen alongside all of our other research and educational responsibilities.  

[Happy Belly Bioinformatics](https://astrobiomike.github.io/) is a website designed and maintained around the goal of helping biologists be able to learn and apply bioinformatics as needed in their research. It has been directly born out of my personal experience of, like many of us these days, being a biologist thrust into the computational world with little or no guidance/advice/resources in my surrounding lab/department/environment. The underlying educational philosophy of the site is that learning a small set of baseline, fundamental skills and concepts that will directly aid in your research is what is important – rather than, for example, sitting down and trying to master a specific programming language without any direct application to your work. When just starting out, it can be very hard to know where to look and where to begin.

[Happy Belly Bioinformatics](https://astrobiomike.github.io/) provides a resource that can be used for both group educational activities such as workshop-type environments as well as by any individual on their own. Primary areas currently covered include: 1) a crash course in working at a [Unix-like command line](https://astrobiomike.github.io/unix/unix-intro) with 5 separate modules ending with [6 commands worth knowing about](https://astrobiomike.github.io/unix/six-glorious-commands) and a primer on [variables and for loops](https://astrobiomike.github.io/unix/for-loops); 2) working in R (e.g. [an introduction to R](https://astrobiomike.github.io/R/basics) and a more detailed look at [indexing in R](https://astrobiomike.github.io/R/more_indexing)); 3) a [comprehensive amplicon/marker-gene tutorial](https://astrobiomike.github.io/amplicon/dada2_workflow_ex) that covers a full workflow from processing through to examples of many standard, initial analyses; and 4) an introduction to some genomics work (e.g. walking through a [*de novo* genome assembly and assessment](https://astrobiomike.github.io/genomics/de_novo_assembly)) and metagenomics (e.g. [recovering **M**etagenome-**A**ssembled **G**enomes, MAGs](https://astrobiomike.github.io/genomics/metagen_anvio)).  

[Happy Belly Bioinformatics](https://astrobiomike.github.io/) helps researchers bridge the gap from having little or no current experience to being comfortable with a baseline set of fundamental skills through comprehensive, heavily annotated, and self-contained tutorials. This equips researchers to then develop their sphere of capabilities in any and all directions as needed, without there any longer being a technical bottleneck on their science.  


